### PR TITLE
[euvr] Separate Billing Payment/History APIs

### PR DESCRIPTION
**Pull request by vincentsalucci**
## Type of change
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Code changes
- **AccountsController.cs**: Added `billing-history` and `billing-payment` API endpoints
- **BillingHistoryResponseModel.cs**: Created response model for housing `Invoices` and `Transactions`
- **BillingPaymentResponseModel.cs**: Created response model for housing `Balance` and `PaymentSource`
- **IPaymentService.cs**: Added new methods to interface: `GetBillingHistoryAsync` and `GetBillingBalanceAndSourceAsync`
- **StripePaymentService.cs**: Cleaned up existing `GetBillingAsync` method by breaking out different properties into functions that can be used independently. Created new methods `GetBillingHistoryAsync` and `GetBillingBalanceAndSourceAsync` for grabbing only the necessary information.

## Testing requirements
- Test new tabs available in `Settings -> Subscription`
- Test `Organization Settings -> Billing` as that will still use the `GetBilling` API that's in place (backwards-compat)

## Before you submit
- [X] I have checked for formatting errors (`dotnet tool run dotnet-format --check`) (required)
- [ ] If making database changes - I have also updated Entity Framework queries and/or migrations
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/Core/Services/Implementations/StripePaymentService.cs [hunk 1]
```diff
@@ -1707,5 +1658,115 @@ private async Task VerifyAppleReceiptNotInUseAsync(string receiptOriginalTransac
...
+                Stripe.CustomerGetOptions customerOptions = null;
+                if (expandForPayments)
+                {
+                    customerOptions = new Stripe.CustomerGetOptions();
+                    customerOptions.AddExpand("default_source");
+                    customerOptions.AddExpand("invoice_settings.default_payment_method");
+                }
+                customer = await _stripeAdapter.CustomerGetAsync(gatewayCustomerId, customerOptions);
+            }
+            catch (Stripe.StripeException) { }
```
Comment (Hinton) [inline src/Core/Services/Implementations/StripePaymentService.cs:1732]: Should we provide an actual exception here?

File: src/Core/Services/Implementations/StripePaymentService.cs [hunk 2]
```diff
@@ -1707,5 +1658,115 @@ private async Task VerifyAppleReceiptNotInUseAsync(string receiptOriginalTransac
...
+
+            Stripe.Customer customer = null;
+            try
+            {
+                Stripe.CustomerGetOptions customerOptions = null;
+                if (expandForPayments)
+                {
+                    customerOptions = new Stripe.CustomerGetOptions();
+                    customerOptions.AddExpand("default_source");
+                    customerOptions.AddExpand("invoice_settings.default_payment_method");
```
Comment (Hinton) [inline src/Core/Services/Implementations/StripePaymentService.cs:1728]: I wonder if we could extract these to a separate method, and have `Stripe.CustomerGetOptions` as a method argument instead. Would simplify it a fair bit.

```c#
var customer = GetCustomerAsync(id, InvoicePaymentMethod())
```

**Review by Hinton — COMMENTED** [review]
Do we want to target this to master?

File: src/Api/Controllers/AccountsController.cs [hunk 3]
```diff
@@ -640,6 +640,34 @@ public async Task<BillingResponseModel> GetBilling()
             return new BillingResponseModel(billingInfo);
         }
```
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Api/Controllers/AccountsController.cs:641]: We should only be fetching it from web which is a place that gets released in sync with server.

[the remark this answers is not in the document]
Comment (vincentsalucci) [reply src/Api/Controllers/AccountsController.cs:641]: While that generally is the case, I feel more comfortable not committing breaking changes to the API at the initial release.

File: src/Api/Controllers/AccountsController.cs [hunk 4]
```diff
@@ -640,6 +640,34 @@ public async Task<BillingResponseModel> GetBilling()
...
+            var user = await _userService.GetUserByPrincipalAsync(User);
+            if (user == null)
+            {
+                throw new UnauthorizedAccessException();
+            }
+
+            var billingInfo = await _paymentService.GetBillingBalanceAndSourceAsync(user);
+            return new BillingPaymentResponseModel(billingInfo);
+        }
+
```
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply src/Api/Controllers/AccountsController.cs:670]: Indifferent. I know the `accounts` controller is bloated. Would it make sense to create a new controller for two calls (even though they are directly related to a user's account) ?

File: src/Core/Services/Implementations/StripePaymentService.cs (hunk 1 above)
Comment (vincentsalucci) [reply src/Core/Services/Implementations/StripePaymentService.cs:1732]: Carryover from original implementation. I'm unsure of its impact in regards to frequency.

File: src/Api/Controllers/AccountsController.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Api/Controllers/AccountsController.cs:670]: I think so. Got to start somewhere.

[the remark this answers is not in the document]
Comment (vincentsalucci) [reply src/Api/Controllers/AccountsController.cs]: Created!

File: src/Core/Services/Implementations/StripePaymentService.cs (hunk 2 above)
Comment (vincentsalucci) [reply src/Core/Services/Implementations/StripePaymentService.cs]: Updated to simplify the method as suggested.

File: src/Api/Controllers/AccountsController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Api/Controllers/AccountsController.cs]: Do we have a tech debt to remove this before the next release?

**Review by Hinton — APPROVED** [review]

**Comment by Hinton** [comment]
Added hold label to ensure we don't merge this before the code freeze elapses.

**Comment by vincentsalucci** [comment]
`rc` branches were cut 4/4/2022. Removing `hold` tag.

File: src/Api/Controllers/AccountsController.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vincentsalucci) [reply src/Api/Controllers/AccountsController.cs]: Created tech debt item and self-assigned for a future release. 👍

**Pull request merged**